cedar-lean/Cedar/Thm/Data/Map.lean

```diff
@@ -164,31 +372,303 @@ theorem mapOnValues_contains {α β γ} [LT α] [DecidableLT α] [DecidableEq α
+      <;> simp only [Option.pure_def, Option.bind_eq_bind, Option.bind_eq_none] at h₃
+      <;> exact h₃ tl' h₅
+    case some mtl' =>
+      simp only [h₁, Option.pure_def, Option.bind_eq_bind, Option.bind_eq_some, Option.some.injEq,
+        List.mapM_cons, mk.injEq, exists_eq_right, List.cons.injEq, exists_eq_right_right,
+        Prod.mk.injEq, true_and] at *
+      apply And.intro h₂
+      replace ⟨tl', h₃, h₄⟩ := h₃
+      subst mtl'
+      simp [h₃]
+
+theorem mapMOnValues_some_implies_forall₂ [LT α] [DecidableLT α] {f : β → Option γ} {m₁ : Map α β} {m₂ : Map α γ} :
+  m₁.mapMOnValues f = some m₂ →
+  List.Forall₂ (λ kv₁ kv₂ => kv₁.fst = kv₂.fst ∧ f kv₁.snd = some kv₂.snd) m₁.kvs m₂.kvs
+:= by
+  unfold mapMOnValues kvs
+  intro h₁
+  simp only [Option.pure_def, Option.bind_eq_bind, Option.bind_eq_some, Option.some.injEq] at h₁
+  replace ⟨x, h₁, h₂⟩ := h₁
+  subst h₂
+  replace h₁ := List.mapM_some_iff_forall₂.mp h₁
+  simp only
+  apply List.Forall₂.imp _ h₁
+  intro (k, v) (k', v') h₂
+  simp only [Option.bind_eq_some, Option.some.injEq, Prod.mk.injEq, exists_eq_right_right] at h₂
+  replace ⟨h₂, h₂'⟩ := h₂
+  subst k'
+  simp only [true_and]
+  exact h₂
+
+theorem mapMOnValues_some_implies_all_some {α : Type 0} [LT α] [DecidableLT α] {f : β → Option γ} {m₁ : Map α β} {m₂ : Map α γ} :
+  m₁.mapMOnValues f = some m₂ →
+  ∀ kv ∈ m₁.kvs, ∃ v, (kv.fst, v) ∈ m₂.kvs ∧ f kv.snd = some v
+:= by
+  intro h₁
+  replace h₁ := List.forall₂_implies_all_left (mapMOnValues_some_implies_forall₂ h₁)
+  intro (k, v) h₂
+  replace ⟨(k', v'), h₁, h₃, h₄⟩ := h₁ (k, v) h₂
+  simp only at *
+  subst k'
+  exists v'
+
+/--
+  alternate proof of `mapMOnValues_some_implies_all_some`, which instead of
+  relying on `mapMOnValues_some_implies_forall₂`, relies on
+  `List.mapM_some_implies_all_some`.  Which do we prefer?
+-/
+theorem mapMOnValues_some_implies_all_some_alt_proof [LT α] [DecidableLT α] {f : β → Option γ} {m₁ : Map α β} {m₂ : Map α γ} :
+  m₁.mapMOnValues f = some m₂ →
+  ∀ kv ∈ m₁.kvs, ∃ v, (kv.fst, v) ∈ m₂.kvs ∧ f kv.snd = some v
+:= by
+  unfold mapMOnValues
+  intro h₁ kv h₂
+  cases h₃ : m₁.kvs.mapM (λ x => match x with | (k, v) => do let v' ← f v ; pure (k, v'))
+  <;> rw [h₃] at h₁
+  <;> simp only [Option.pure_def, Option.bind_some_fun, Option.bind_none_fun, Option.some.injEq] at h₁
+  case some ags =>
+    subst h₁
+    have (a, b) := kv ; clear kv
+    simp only
+    replace h₃ := List.mapM_some_implies_all_some h₃
+    replace ⟨(a', g), h₃, h₄⟩ := h₃ (a, b) h₂
+    simp only [Option.pure_def, Option.bind_eq_bind, Option.bind_eq_some, Option.some.injEq,
+      Prod.mk.injEq, exists_eq_right_right] at h₄
+    replace ⟨h₄, h₄'⟩ := h₄
+    subst a'
+    exists g
+
+theorem mapMOnValues_some_implies_all_from_some [LT α] [DecidableLT α] {f : β → Option γ} {m₁ : Map α β} {m₂ : Map α γ} :
+  m₁.mapMOnValues f = some m₂ →
+  ∀ kv ∈ m₂.kvs, ∃ v, (kv.fst, v) ∈ m₁.kvs ∧ f v = kv.snd
```

Comment:
Pretty similar in form to `mapMOnValues_some_implies_all_some`. Do you really need both? Is there a way to combine them into one statement?